lib/providers/metrics-provider.js

```diff
@@ -1,25 +1,461 @@
+    var averagedAggregate = {};
+
+    for (var dataKey in data) {
+      averagedAggregate[dataKey] = {};
+
+      for (var dataMetricKey in data[dataKey]) {
+        averagedAggregate[dataKey][dataMetricKey] = 0;
+
+        // the empty aggregate is created with zero values
+        if (!aggregate) {
+          continue;
+        }
+
+        // you can compute an average of a set of numbers two ways
+        // first, you can add all the numbers together and then divide by the count
+        // second, you call divide each number by the count and add the quotients
+        // the first method is more accurate, however you can overflow an accumulator
+        // and result with NaN
+        // the second method is employed here to ensure no overflows
+        for (var index = 0; index < aggregate[dataKey][dataMetricKey].values.length; index++) {
+          averagedAggregate[dataKey][dataMetricKey] +=
+            aggregate[dataKey][dataMetricKey].values[index] /
+            aggregate[dataKey][dataMetricKey].values.length;
+        }
+
+        // truncate the number to one decimal point
+        averagedAggregate[dataKey][dataMetricKey] =
+          +averagedAggregate[dataKey][dataMetricKey].toFixed(1);
+      }
+    }
+
+    return averagedAggregate;
+  };
+
+/**
+ * Given a metric data template, and a metric data point, accumulate into the
+ * aggregator.
+ *
+ * @param {Object} data
+ * The metric data template received.
+ *
+ * @param {Object} metric
+ * The metric data point received.
+ *
+ * @param {Object} aggregate
+ * The aggregator object.
+ *
+ * @returns {void}
+ */
+var accumulateAggregate =
+  function accumulateAggregate(data, metric, aggregate) {
+    for (var dataKey in data) {
+      for (var dataMetricKey in data[dataKey]) {
+        aggregate[dataKey][dataMetricKey].values.push(metric[dataKey][dataMetricKey]);
+      }
+    }
+  };
+
+/**
+ * This is the constructor for the MetricsProvider
+ *
+ * @param {Object} screen
+ * The blessed screen object.
+ *
+ * @returns {void}
+ */
+var MetricsProvider =
+  function MetricsProvider(screen) {
+    EventEmitter.call(this);
+
+    this._metrics = [];
 
+    // construct the aggregation container
+    this._aggregation = {};
+    for (var index = 0; index < config.AGGREGATE_TIME_BUCKETS.length; index++) {
+      this._aggregation[config.AGGREGATE_TIME_BUCKETS[index]] = {
+        lastIndex: 0,
+        data: []
+      };
+    }
+
+    this.aggregationLevels = Object.keys(this._aggregation);
+    this.currentAggregateZoom = -1;
+    this.currentAggregateZoomKey = this.aggregationLevels[this.currentAggregateZoom];
+    this.minimumAggregationInterval = +this.aggregationLevels[0];
+
+    this._startTime = Date.now();
+    this._lastAggregation = Date.now();
+
+    screen.on("metrics", this._onMetrics.bind(this));
+
+    screen.on("zoomAggregate", function zoomAggregate(zoom) {
+      // apply zoom delta and check for boundaries
+      this.currentAggregateZoom += zoom;
```

Comment:
Could we just use something like lodash.clamp (https://lodash.com/docs/4.17.4#clamp) to enforce bounds?